### ISSUE TITLE
System.OutOfMemoryException when Rendering Plot in Winforms

### ISSUE THREAD

**Issue by MEGA-GAGA** [oxyplot/oxyplot#2037]
### Steps to reproduce

1. **Actually there is no exact way to reproduce**
2. **Sometimes the application runs many hours and sometimes it throws exception after an hour**

Platform:
.NET version:
**.NET Framework 4.7.2**
**Windows 10 and Windows 11**
**Winforms**

### Expected behaviour

**System.OutOfMemoryException should not occur**

### Actual behaviour

Sometimes the application runs many hours and sometimes it throws an exception after an hour. Please see attached screenshot.

The behaviour with our exe looks nearly like in #1264

But I do not exactly understand the workaround "Use PlotModel" which is suggested in #1264. It means not to use an explicit ItemsSource and put the points directly into LineSeries?

Can someone please explain.

How can I avoid the System.OutOfMemoryException when using oxyplot like I do?

Number of points is definitely not the problem.

Regardless of this case I did a test application with hundred times more points and there was no problem.

![Snag_3fc6ef5](https://github.com/oxyplot/oxyplot/assets/144467942/2e8fda11-66a9-49c7-93a9-00c87a827654)

Thank You!

**Comment by VisualMelon** [comment]
The workaround mentioned in #1264 is specific to WPF and not applicable to WinForms.

Can you describe any more of what you are doing? Are you using ItemsSource or Points and how are you updating the series? Can you get a memory profile/snapshot so we can see what is being initialised? Which version of OxyPlot are you using? The GDI renderer used to leak brushes and all sorts, but that was years ago, so if you're using a recent version there shouldn't be anything like that going on. Note that while it may be OxyPlot that dies each time, it may be that another component is leaking memory, and it's just that OxyPlot has a high memory churn when rendering that causes the OOM.

**Comment by MEGA-GAGA** [comment]
First I would like to state that after reading the post I would assume it is a problem with the pc or with the windows operating system the screenshot above was taken from.

The application reads data from a sensor via serial interface every 66 ms and plots them with oxyplot in two plots. Normally it looks like shown below in **Picture 1**.

The user can switch the time axis range between 30 s up to 30 min. Therefore I keep data of the last 30 min in the background which means to have two times 26828 items in lists.

As one can see there are two PlotViews with one LineSeries each. Each LineSeries has an ItemSource set. The ItemSource is just a .NET typed List of type PointType. PointType is the same as DataPoint in ItemsSourceExamples.cs. All in all, it is very similar to an oxyplot example.

When there are points older then 30 min I remove them from the list. Unfortunately the problem was reported by a user who is a little further away and not a software developer and does not have a development machine. The screenshot above is from this machine and he and me communicate by email.

Now I have done some research on my development computer and cannot find that there is a memory leak problem or anything like that at all. I attached a screenshot from running the application with diagnostic tool. The behaviour is as expected. Memory consumption will rise in the first 30 min and then stay stable. At id 6 I put the application into fullscreen on a 4k monitor and back to normal size. Values are below any limits for an actual pc. I cannot see a problem.

When I receive new data I put a point into the LineSeries, calculate the x-Axis and the y-Axis and call ResetAllAxes() and InvalidatePlot(). (y-Axis is recalculated because there is an auto range mode). This means I call ResetAxes() and InvalidatePlot() every 66 ms.

The exception above comes from inside ExtractNextContiguousLineSegment in file LineSeries.cs when trying to increase the size of a list. The capacity of the list is increased to twice the current capacity: [List<T>.EnsureCapacity(Int32) Method](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.ensurecapacity?view=net-7.0)

The exception ocurs randomly on the users pc after one hour. The rest of the application continues to run, but this is probably because the exception freed up memory.

First of all it is important for me to understand what is going wrong on the users pc. 

And second if the reason for the exception can not be solved it should not show up the way it shows up. Because in this way it might be understood as an oxyplot problem. Later there should be a more user friendly message implemented that there could be no rendering because of memory issues, deavtivate the plot and request to restart e.g.

The application will be shipped to different customers and depending on there pc the issue might arise again and I will have to handle that.

For now I have following ideas to investigate with the user:

 - Examination of overall memory consumption of the users pc when running
 - Check if windows is up to date with drivers GDI etc. (it is windows 10 but I dont know which version exactly)
 - Making a debug build which logs Private Memory, Heap consumption and free memory to a file (because there is no development environment on the user pc).

Oxyplot is from 2022 when we started with the project. I will run and test with the actual version of course.

Will using .NET 6 have an advantage over .NET Framework 4.7.2?

There might be a simpler rendering mode with less resource requirements?

If anyone has an interesting idea worth following please let me know.

**Picture 1**: Normal working application

![Snag_9bb7710](https://github.com/oxyplot/oxyplot/assets/144467942/d2994de1-e9e5-44a0-aad8-ecc7d4255dbb)

**Picture 2**: Diagnostic Tool

![Snag_9e236ae](https://github.com/oxyplot/oxyplot/assets/144467942/390694dc-01c4-4c31-86f4-3289d330dbc6)

**Comment by VisualMelon** [comment]
Thanks for the additional information: definitely sound like we'll need diagnostics from the user's machine to track this down if it's not been observed elsewhere. You could use PerfView to get a couple of memory snapshots (e.g. before and after it dies): it's lightweight compared to say VS.

I wouldn't expect much difference between .NET 6 and netframework 4.7, but I really don't know: worth a try I'd say because it's an easy substitution and I'm guessing the user doesn't really care what's wrong as long as it works.

I'd be shocked if this was a bug in the GDI because it's so ancient. There isn't a simpler rendering mode or anything like that in oxyplot,

**Comment by MEGA-GAGA** [comment]
The feedback from the user is that the issue no more appeared. It is an actual win 11 pc with 32 GB RAM. I assume that he had worked more or less every day with the application since the last days. (Maybe a windows update was running when the issue occured, but this is only a guess.)

I have done some research here on a Dell Optiplex XE with 8 GB RAM. I have tortured the PC with a RAM allocation tool up to 99 % located RAM. I did this many times. Always the PC stood still or something else happened and it crashed.

But never showed up the issue above!

From my point of view, this can be considered a very rare case and ticket can be closed.

If I will be able to reproduce I will come back.
